name: CI orchestrator

on:
  pull_request:
    branches:
      - main
  merge_group:
    types: [checks_requested]

concurrency:
  group: ${{ github.workflow }}-${{ github.ref }}
  cancel-in-progress: true

permissions:
  contents: write
  checks: write
  pull-requests: write

jobs:
  changes:
    uses: ./.github/workflows/detect-changes.yml

  #
  # Reusable CI
  #
  ci_reusable:
    name: Continuous Integration
    uses: ./.github/workflows/reusable-ci.yml
    with:
      packages_directory: source
      pytest_addopts: --ignore-glob 'tests/**/subsystem_tests/**'
    secrets:
      dh3serviceaccount_privatekey: ${{ secrets.dh3serviceaccount_privatekey }}

  #
  # Dotnet CI
  #
  ci_dotnet:
    needs: changes
    if: ${{ needs.changes.outputs.dotnet == 'true' }}
    uses: ./.github/workflows/ci-dotnet.yml

  #
  # Branch policy status check
  #
  allow_merge_ci_orchestrator:
    runs-on: ubuntu-24.04
    needs: [ci_reusable]
    if: |
      always()
    steps:
      - name: Verify if merge is allowed
        run: |
          echo "${{ toJSON(needs) }}"
          if [[ ${{ contains(needs.*.result, 'failure') || contains(needs.*.result, 'cancelled') }} = true ]]; then
              echo "Failed"
              exit 1
          fi
